File: medium/1461/main.rs
impl Solution {
    pub fn has_all_codes(s: String, k: i32) -> bool {
        let len = s.len();
        let mut possible = 1 << k;
        let tmp = possible;
        let mut hash = vec![false; possible];
        let mut val = 0;

        for i in 0..len {
            val = ((val << 1) & !tmp) + (s.as_bytes()[i] - b'0') as usize;
            if i+1 >= k as usize && hash[val] == false {
                hash[val] = true;
                possible -= 1;
                if possible == 0 {
                    return true;
                }
            }
        }
        
        false
    }
}

File: medium/1461/main.c
bool hasAllCodes(char* s, int k) {
  // there are 2^k possiable solutions.
  int possible = 1 << k;
  bool hash[possible];
  int len = strlen(s);
  int val = 0;

  memset(hash, 0, sizeof(hash));

  for (int i = 0, tmp = possible; i < len; i++) {
    val = ((val << 1) & ~tmp) + s[i] - '0';
    if (i+1 >= k && hash[val] == false) {
      hash[val] = true;
      possible--;
      if (possible == 0) return true;
    }
  }
    
  return false;

}
